src/Control/Carrier/ContainerRegistryApi/Authorization.hs

```diff
@@ -76,18 +81,40 @@ mkRequest ::
 mkRequest manager registryCred accepts req = do
   token <- getToken =<< ask
   token' <- getAuthToken registryCred req manager token
-  logHttp (applyContentType accepts $ applyBearAuth token' req) manager
+  logHttp (applyContentType accepts $ applyAuthToken token' req) manager
   where
     applyContentType :: Maybe [Text] -> Request -> Request
     applyContentType c r = case c of
       Nothing -> req
       Just c' -> r `acceptsContentType` (Text.intercalate ", " c')
 
--- | Adds 'Authorization: Bearer <>', if 'Just token', otherwise id.
-applyBearAuth :: Maybe Text -> Request -> Request
-applyBearAuth t r = case t of
-  Nothing -> r
-  Just t' -> applyBearerAuth (encodeUtf8 t') r
+-- | Adds 'Authorization' header if token is provided otherwise id
+applyAuthToken :: Maybe AuthToken -> Request -> Request
+applyAuthToken Nothing r = r
+applyAuthToken (Just (BasicAuthToken user pass)) r =
+  stripAuthHeaderOnRedirect $ applyBasicAuth (encodeUtf8 user) (encodeUtf8 pass) r
+applyAuthToken (Just (BearerAuthToken token)) r =
+  stripAuthHeaderOnRedirect $ applyBearerAuth (encodeUtf8 token) r
+
+-- If we don't strip auth headers, on redirect, depending on how
+-- blobs/manifest are retrieved cloud vendor may throw 'Bad Request' error.
+stripAuthHeaderOnRedirect :: Request -> Request
```

Comment:
In practice, I've noticed sometimes cloud vendors are doing redirects for blobs, in which, if Auth headers are passed, it will cause 400 Bad Request Errors. This ultimately leads to `Z_DATA_ERROR (-3)` from zlib.

Reply:
I have not been able to reproduce this conditions for other vendors - quay, docker, gitub, gcr.